### ISSUE TITLE
Null Reference Exception occurred on mouse move the dropdown of Inherit combobox in NETCore

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/winforms#5154]
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/t/Null-Reference-Exception-occurred-on-mou/1459552)._

---
Hi Team,

we have create a control on inherit to Combobox and create a CreateAccessibilityInstance() in NETCORE 50 Frame work , when open the dropdown to mouse over the item or touch a item exception occured from the base. WndProc().we have removed a CreateAccessibilityInstance() override a method its working fine. we have attached a simple sample in below, Please share the Details in ASAP.
 
**Note**  - Issue reproduced in Netcore framework only

Sample - 
[Mscombobox_Exception.zip] (https://aka.ms/dc/file?name=B004e361376244c318cb5ab0a216938b7637601182536887178_Mscombobox_Exception.zip&tid=004e361376244c318cb5ab0a216938b7637601182536887178)

**Exception Details**

System.NullReferenceException
  HResult=0x80004003
  Message=Object reference not set to an instance of an object.
  Source=System.Windows.Forms
  StackTrace:
   at System.Windows.Forms.ComboBox.ComboBoxChildNativeWindow.WndProc(Message& m)
   at System.Windows.Forms.NativeWindow.DebuggableCallback(IntPtr hWnd, Int32 msg, IntPtr wparam, IntPtr lparam)
   at System.Windows.Forms.UnsafeNativeMethods.DispatchMessageW(MSG& msg)
   at System.Windows.Forms.Application.ComponentManager.System.Windows.Forms.UnsafeNativeMethods.IMsoComponentManager.FPushMessageLoop(IntPtr dwComponentID, Int32 reason, Int32 pvLoopData)
   at System.Windows.Forms.Application.ThreadContext.RunMessageLoopInner(Int32 reason, ApplicationContext context)
   at System.Windows.Forms.Application.ThreadContext.RunMessageLoop(Int32 reason, ApplicationContext context)
   at System.Windows.Forms.Application.Run(Form mainForm)
   at WindowsFormsApp73.Program.Main() in C:\Users\SulthanAmanu\source\repos\WindowsFormsApp73\Program.cs:line 20
 
Regards,
Sulthan amanu

---
#### Feedback Bot on 6/24/2021, 04:10 PM: 

<p>We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>

---
### Original Solutions
(no solutions)

**Comment by John-Qiao** [comment]
* .NET Core Version: 
.NET SDK 6.0.100-preview.7.21324.6

* Have you experienced this same bug with .NET Framework?: 
Yes, this issue occurs in .NET Framework 4.8 too. But it does not occur in .NET Framework 4.7.2 or earlier versions.

**Problem description:**

The exception occurs when move mouse over the item in inherited ComboBox control.
![Core-6 0](https://user-images.githubusercontent.com/45864985/123382167-338b3700-d5c4-11eb-97b8-7b8e9bc06814.png)
`System.InvalidCastException`
`  HResult=0x80004002`
`  Message=Unable to cast object of type 'PastelessComboBoxAccessibility' to type 'ComboBoxAccessibleObject'.`
`  Source=System.Private.CoreLib`
`  StackTrace:`
`   at System.Runtime.CompilerServices.CastHelpers.ChkCastAny(Void* toTypeHnd, Object obj)`
`   at System.Windows.Forms.ComboBox.ObjectCollection..ctor(ComboBox owner)`
`   at System.Windows.Forms.ComboBox.get_Items()`
`   at TestApp_Core.Form1..ctor() in C:\Users\username\Desktop\0625\TestApp-Core\TestApp-Core\Form1.cs:line 21`
`   at TestApp_Core.Program.Main() in C:\Users\username\Desktop\0625\TestApp-Core\TestApp-Core\Program.cs:line 20`

**Expected behavior:**

The item can be selected successfully in inherited ComboBox control, like it works when targeting project to .NET Framework 4.7.2. [TestApp-Framework.zip](https://github.com/dotnet/winforms/files/6714094/TestApp-Framework.zip)
![Framework-4 7 2](https://user-images.githubusercontent.com/45864985/123382142-2c642900-d5c4-11eb-9615-4089d37103be.png)

**Minimal repro:**

1. Open the attached [TestApp-Core.zip](https://github.com/dotnet/winforms/files/6713986/TestApp-Core.zip) project in VS and run it.
2. Expand the inherited ComboBox control, and try to move mouse over one item in it.

**Mo Info:**

1. This is a regression issue, it does not occur in .NET Framework 4.7.2 or earlier versions, but occurs in .NET Framework 4.8 as below.
![Framework-4 8](https://user-images.githubusercontent.com/45864985/123381565-82849c80-d5c3-11eb-86ce-79963a43dc32.png)
`System.NullReferenceException`
`  HResult=0x80004003`
`  Message=Object reference not set to an instance of an object.`
`  Source=System.Windows.Forms`
`  StackTrace:`
`   at System.Windows.Forms.ComboBox.ComboBoxChildNativeWindow.WndProc(Message& m)`
`   at System.Windows.Forms.NativeWindow.DebuggableCallback(IntPtr hWnd, Int32 msg, IntPtr wparam, IntPtr lparam)`
`   at System.Windows.Forms.UnsafeNativeMethods.DispatchMessageW(MSG& msg)`
`   at` `System.Windows.Forms.Application.ComponentManager.System.Windows.Forms.UnsafeNativeMethods.IMsoComponentManager.FPushMessageLoop(IntPtr dwComponentID, Int32 reason, Int32 pvLoopData)`
`   at System.Windows.Forms.Application.ThreadContext.RunMessageLoopInner(Int32 reason, ApplicationContext context)`
`   at System.Windows.Forms.Application.ThreadContext.RunMessageLoop(Int32 reason, ApplicationContext context)`
`   at System.Windows.Forms.Application.Run(Form mainForm)`
`   at TestApp_Framework.Program.Main() in C:\Users\username\Desktop\0625\TestApp-Framework\TestApp-`
`Framework\Program.cs:line 19`

2. This issue occurs in .NET Core project too, but the exception results are different between .NET 6.0 and .NET 5.0/NET Core 3.1. In .NET 5.0 and NET Core 3.1, the exception is similar with .NET Framework 4.8 testing result, please check below result in .NET 5.0 and NET Core 3.1.
![Core-5 0](https://user-images.githubusercontent.com/45864985/123380928-a2679080-d5c2-11eb-9f43-1779ac78d62d.png)
`System.NullReferenceException`
 ` HResult=0x80004003`
 ` Message=Object reference not set to an instance of an object.`
 ` Source=System.Windows.Forms`
 ` StackTrace:`
 `  at System.Windows.Forms.ComboBox.ComboBoxChildNativeWindow.WndProc(Message& m)`
`   at System.Windows.Forms.NativeWindow.Callback(IntPtr hWnd, WM msg, IntPtr wparam, IntPtr lparam)`
 `  at Interop.User32.DispatchMessageW(MSG& msg)`
 `  at` `System.Windows.Forms.Application.ComponentManager.Interop.Mso.IMsoComponentManager.FPushMessageLoop(UIntPtr` `dwComponentID, msoloop uReason, Void* pvLoopData)`
`   at System.Windows.Forms.Application.ThreadContext.RunMessageLoopInner(msoloop reason, ApplicationContext context)`
 `  at System.Windows.Forms.Application.ThreadContext.RunMessageLoop(msoloop reason, ApplicationContext context)`
 `  at System.Windows.Forms.Application.Run(Form mainForm)`
`   at TestApp_Core.Program.Main() in C:\Users\username\Desktop\0625\TestApp-Core\TestApp-Core\Program.cs:line 20`
`  This exception was originally thrown at this call stack:`
  `  [External Code]`
  `  TestApp_Core.Program.Main() in Program.cs`

**Comment by RussKie** [comment]
@vladimir-krestov @SergeySmirnov-Akvelon please investigate

**Comment by vladimir-krestov** [comment]
Please see the following code:
https://github.com/dotnet/winforms/blob/8c12a8276fa2d9e6ced90a92eae3fd6badb4fceb/src/System.Windows.Forms/src/System/Windows/Forms/ComboBox.ComboBoxChildNativeWindow.cs#L44-L47

In the customer code `_owner.AccessibilityObject` is `PastelessComboBoxAccessibility`, that is not "ComboBoxAccessibleObject".

There is no way to inherit a new custom class from `ComboBoxAccessibleObject`, because it is internal. Anyway, there is no way to implement a correct custom accessible object for `ComboBox`, because to do that you need to override several important methods and properties, that are internal (eg. `GetPropertyValue`, `FragmentRoot`, `FragmentNavigate`). You won't build a correct accessible ojbect of a complex control without them.
To return 4.7.2 behavior you need to set `SupportsUiaProvider` as `false`, but it's impossible from a user code, because this property is internal.
So, implementing a custom accessible object for a complex control (eg. ComboBox) is not a good idea.

Actually, I can rework this logic to avoid the exception throwing, but it won't resolve feature issues.
@RussKie, @dreddy-work, should I rework the logic even if it's impossible to build a correct accessible object next?

**Comment by dreddy-work** [comment]
> So, implementing a custom accessible object for a complex control (eg. ComboBox) is not a good idea.

But we do allow creation of CustomeAccessible object. correct? Setting `SupportsUiaProvider ` to `false` help achieve that? How do we support applications being migrated? @Tanya-Solyanik 
 
> Actually, I can rework this logic to avoid the exception throwing, but it won't resolve feature issues.
> @RussKie, @dreddy-work, should I rework the logic even if it's impossible to build a correct accessible object next?

We should do that meanwhile.

**Comment by RussKie** [comment]
> Please see the following code:
> https://github.com/dotnet/winforms/blob/8c12a8276fa2d9e6ced90a92eae3fd6badb4fceb/src/System.Windows.Forms/src/System/Windows/Forms/ComboBox.ComboBoxChildNativeWindow.cs#L44-L47
> 
> In the customer code `_owner.AccessibilityObject` is `PastelessComboBoxAccessibility`, that is not "ComboBoxAccessibleObject".
> 
> There

This code doesn't exist in .NET Framework, so it is something we added later.
Perhaps the code can be expressed as follows:
```cs
                    case WM.MOUSEMOVE:
                        if (_childWindowType == ChildWindowType.DropDownList)
                        {
                            // Need to track the selection change via mouse over to
                            // raise focus changed event for the items. Monitoring
                            // item change in setters does not guarantee that focus
                            // is properly announced.
                            object before = _owner.SelectedItem;
                            DefWndProc(ref m);
                            object after = _owner.SelectedItem;
                            if (before != after && _owner.AccessibilityObject is ComboBoxAccessibleObject comboBoxAccessibleObject)
                            {
                                comboBoxAccessibleObject.SetComboBoxItemFocus();
                            }
                        }
                        else
                        {
                            _owner.ChildWndProc(ref m);
                        }
```

**Comment by vladimir-krestov** [comment]
@dreddy-work, 
> But we do allow creation of CustomeAccessible object. correct?
Yes, but it won't be successful.
> Setting `SupportsUiaProvider` to `false` help achieve that?
Yes, but a user can't do it even using reflection.

I have fixed the case for ComboBox.
I tried to find any places with the same issue using " as (\w|.)*AccessibleObject" regular expression.
There are no any incorrect usings of `as`, there are internal objects only.

**Comment by John-Qiao** [comment]
Verified with .NET 6.0.100-rc.1.21368.3 from main branch, this issue has been fixed in this .NET 6.0 SDK that items can be selected in Inherit combobox successfully. 
![TestResult](https://user-images.githubusercontent.com/45864985/126124347-0dde3e44-26fd-4166-9b1c-4bbe786da0ab.gif)

**Comment by Olina-Zhang** [comment]
Verified with the .Net 6.0 Preview 7 release build: 6.0.100-Preview.7.21369.19, it is fixed as above result.
